models/rnnt.py

```diff
@@ -0,0 +1,202 @@
+
+    # encoder
+    for i in range(2):
+      self.encoder.pre_rnn.cells[i].weights_ih.assign(state_dict[f"encoder.pre_rnn.lstm.weight_ih_l{i}"].numpy())
+      self.encoder.pre_rnn.cells[i].weights_hh.assign(state_dict[f"encoder.pre_rnn.lstm.weight_hh_l{i}"].numpy())
+      self.encoder.pre_rnn.cells[i].bias_ih.assign(state_dict[f"encoder.pre_rnn.lstm.bias_ih_l{i}"].numpy())
+      self.encoder.pre_rnn.cells[i].bias_hh.assign(state_dict[f"encoder.pre_rnn.lstm.bias_hh_l{i}"].numpy())
+    for i in range(3):
+      self.encoder.post_rnn.cells[i].weights_ih.assign(state_dict[f"encoder.post_rnn.lstm.weight_ih_l{i}"].numpy())
+      self.encoder.post_rnn.cells[i].weights_hh.assign(state_dict[f"encoder.post_rnn.lstm.weight_hh_l{i}"].numpy())
+      self.encoder.post_rnn.cells[i].bias_ih.assign(state_dict[f"encoder.post_rnn.lstm.bias_ih_l{i}"].numpy())
+      self.encoder.post_rnn.cells[i].bias_hh.assign(state_dict[f"encoder.post_rnn.lstm.bias_hh_l{i}"].numpy())
+
+    # prediction
+    self.prediction.emb.weight.assign(state_dict["prediction.embed.weight"].numpy())
+    for i in range(2):
+      self.prediction.rnn.cells[i].weights_ih.assign(state_dict[f"prediction.dec_rnn.lstm.weight_ih_l{i}"].numpy())
+      self.prediction.rnn.cells[i].weights_hh.assign(state_dict[f"prediction.dec_rnn.lstm.weight_hh_l{i}"].numpy())
+      self.prediction.rnn.cells[i].bias_ih.assign(state_dict[f"prediction.dec_rnn.lstm.bias_ih_l{i}"].numpy())
+      self.prediction.rnn.cells[i].bias_hh.assign(state_dict[f"prediction.dec_rnn.lstm.bias_hh_l{i}"].numpy())
+
+    # joint
+    self.joint.l1.weight.assign(state_dict["joint_net.0.weight"].numpy())
+    self.joint.l1.bias.assign(state_dict["joint_net.0.bias"].numpy())
+    self.joint.l2.weight.assign(state_dict["joint_net.3.weight"].numpy())
+    self.joint.l2.bias.assign(state_dict["joint_net.3.bias"].numpy())
+
+
+class LSTMCell:
+  def __init__(self, input_size, hidden_size):
+    self.hidden_size = hidden_size
+
+    self.weights_ih = Tensor.uniform(hidden_size * 4, input_size)
+    self.bias_ih = Tensor.uniform(hidden_size * 4)
+    self.weights_hh = Tensor.uniform(hidden_size * 4, hidden_size)
+    self.bias_hh = Tensor.uniform(hidden_size * 4)
+
+  def __call__(self, x, h, c):
+    t = (x @ self.weights_ih.T) + self.bias_ih + (h @ self.weights_hh.T) + self.bias_hh
+
+    i, f, g, o = t.chunk(4, 1)
+    i, f, g, o = i.sigmoid(), f.sigmoid(), g.tanh(), o.sigmoid()
+
+    c = (f * c) + (i * g)
+    h = o * c.tanh()
+
+    return h.realize(), c.realize()
+
+
+class LSTM:
+  def __init__(self, input_size, hidden_size, layers, dropout):
+    self.input_size = input_size
+    self.hidden_size = hidden_size
+    self.layers = layers
+    self.dropout = dropout
+
+    self.cells = [LSTMCell(input_size, hidden_size) if i == 0 else LSTMCell(hidden_size, hidden_size) for i in range(layers)]
+
+  def __call__(self, x, h, c):
+    if h is None or c is None:
```

Comment:
Creating tensors is slow.

Why do you have to make copies?

x.assign(<new tensor>) is better than x = <new tensor>

Reply:
the copy here is actually a copy of the list not a copy of the tensors

Reply:
also we need the zero tensors here are this is state that is passed between the layers